src/ContainerInstance.ts

```diff
@@ -333,7 +333,14 @@ export class ContainerInstance {
             // need to be injected, and user can use provided container to get instances he needs
             params.push(this);
 
+            if (type.prototype.OnBefore) {
+                type.prototype.OnBefore.bind(type)();
```

Comment:
It does not seem right to bind a method like this to a Type. Invoking a member method and altering the `this` to be something other than an instance of that Type is naughty.